Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs

```diff
@@ -379,7 +380,12 @@ endRide handle@ServiceHandle {..} rideId req = withLogTag ("rideId-" <> rideId.g
                         else
                           if (isJust updRide.estimatedTollCharges || isJust updRide.tollCharges) && distanceCalculationFailure
                             then (Nothing, Nothing, Just Unsure)
-                            else (updRide.tollCharges, updRide.tollNames, Just Sure)
+                            else
+                              if (isJust updRide.estimatedTollCharges) && (updRide.tollConfidence == Just Unsure || updRide.tollConfidence == Nothing) && (isJust updRide.driverDeviatedToTollRoute)
+                                then (updRide.estimatedTollCharges, updRide.tollNames, Just Sure)
+                                else (updRide.tollCharges, updRide.tollNames, Just Sure)
+
+                _ <- when (isJust updRide.driverDeviatedToTollRoute && updRide.tollConfidence == Just Sure && updRide.tollCharges == Just 0 && isJust updRide.estimatedTollCharges) $ pushToKafka updRide "Toll-estimate-at-end-ride" (show updRide.id)
```

Comment:
- Replace `updRide.tollConfidence` to `tollConfidence`
- In Kafka you have to push this data `data RideIntrerpolationData = RideIntrerpolationData { interpolatedPoints, rideId }` instead of `updRide`
   - interpolatedPoints you would get from this `getInterpolatedPoints ride.driverId`
- Replace `updRide.tollCharges == Just 0` to `maybe True (> 0) updRide.tollCharges`
- Topic name can be `ride-interpolated-waypoints`
- Replace `(show updRide.id)` to `updRide.id.getId`

Reply:
Also, this computation can be forked in seperate thread.